common/bspfile.cc

```diff
@@ -984,6 +984,130 @@ MBSPtoQ2_Models(const dmodelh2_t *dmodelsh2, int nummodels) {
+================
+*/
+static std::vector<uint8_t> CalcPHS(int32_t portalclusters, const uint8_t *visdata, int *visdatasize, int32_t bitofs[][2])
+{
+    const int32_t leafbytes = (portalclusters + 7) >> 3;
+    const int32_t leaflongs = leafbytes / sizeof(long);
+    std::vector<uint8_t> compressed_phs;
+    uint8_t *uncompressed = (uint8_t *) calloc(1, leafbytes);
+    uint8_t *uncompressed_2 = (uint8_t *) calloc(1, leafbytes);
+    uint8_t *compressed = (uint8_t *) calloc(1, leafbytes * 2);
+    uint8_t *uncompressed_orig = (uint8_t *) calloc(1, leafbytes);
+
+    printf ("Building PHS...\n");
+
+    int32_t count = 0;
+    for (int32_t i = 0; i < portalclusters; i++)
+    {
+        const uint8_t *scan = &visdata[bitofs[i][DVIS_PVS]];
+
+        DecompressRow(scan, leafbytes, uncompressed);
+        memset(uncompressed_orig, 0, leafbytes);
+        memcpy(uncompressed_orig, uncompressed, leafbytes);
+
+        scan = uncompressed_orig;
+
+        for (int32_t j = 0; j < leafbytes; j++)
+        {
+            uint8_t bitbyte = scan[j];
+            if (!bitbyte)
+                continue;
+            for (int32_t k = 0; k < 8; k++)
+            {
+                if (! (bitbyte & (1<<k)) )
+                    continue;
+                // OR this pvs row into the phs
+                int32_t index = ((j<<3)+k);
+                if (index >= portalclusters)
+                    Error ("Bad bit in PVS");	// pad bits should be 0
+                const uint8_t *src_compressed = &visdata[bitofs[index][DVIS_PVS]];
+                DecompressRow(src_compressed, leafbytes, uncompressed_2);
+                const long *src = (long *) uncompressed_2;
+                long *dest = (long *) uncompressed;
+                for (int32_t l = 0; l < leaflongs; l++)
+                    ((long *)uncompressed)[l] |= src[l];
+            }
+        }
+        for (int32_t j = 0; j < portalclusters; j++)
+            if (uncompressed[j>>3] & (1<<(j&7)) )
+                count++;
+
+        //
+        // compress the bit string
+        //
+        int32_t j = CompressRow (uncompressed, leafbytes, compressed);
+
+        bitofs[i][DVIS_PHS] = compressed_phs.size();
+
+        compressed_phs.insert(compressed_phs.end(), compressed, compressed + j);
+    }
+
+    free(uncompressed);
+    free(uncompressed_2);
+    free(compressed);
+    free(uncompressed_orig);
+
+    printf ("Average clusters hearable: %i\n", count / portalclusters);
+
+    return compressed_phs;
+}
+
+static dvis_t *
+MBSPtoQ2_Vis(const uint8_t *visdata, int *visdatasize, int numleafs, const mleaf_t *leafs) {
+    int32_t num_clusters = 0;
+    
+    for (int32_t i = 0; i < numleafs; i++) {
+        num_clusters = std::max(num_clusters, leafs[i].cluster + 1);
+    }
+
+    size_t vis_offset = sizeof(dvis_t) + (sizeof(int32_t) * num_clusters * 2);
+    dvis_t *vis = (dvis_t *)calloc(1, vis_offset + *visdatasize);
+
+    vis->numclusters = num_clusters;
+
+    uint8_t test[99999];
```

Comment:
looks like this `test` array can be deleted

Reply:
Oops